job manager: add job states, exceptions

garlick opened the pull request:
This is some cleanup to the job manager, the `flux-job` command, and the libjob API, driven in part by the state and exception descriptions in [RFC 21](https://github.com/flux-framework/rfc/blob/master/spec_21.adoc)

There is now a `flux job raise` command that can raise arbitrary exceptions:
```
Usage: flux-job raise [OPTIONS] id [message ...]
Raise exception for job
  -h, --help             Display this message.
  -s, --severity=N       Set exception severity [0-7] (default=0)
  -t, --type=TYPE        Set exception type (default=cancel)
```
`flux job cancel` is now an alias to `flux job raise`.  Only exceptions with `severity=0` have an effect on the job at this point.

The `flux_job_cancel()` API function is replaced with:
```c
/* Raise an exception for job.
 * Severity is 0-7, with severity=0 causing the job to abort.
 * Note may be NULL or a human readable message.
 */
flux_future_t *flux_job_raise (flux_t *h, flux_jobid_t id,
                               const char *type, int severity,
                               const char *note);
```

Jobs now have states, now reported with a column in `flux job list` output:
```
$ flux job list
JOBID		STATE	USERID	PRI	T_SUBMIT
96552878080	C	5588	16	2019-02-12T22:43:45Z
165490458624	C	5588	16	2019-02-12T22:43:49Z
3035736571904	N	5588	16	2019-02-12T22:46:40Z
```
At the moment jobs are created in `NEW` state, and they transition to `CLEANUP` when you cancel them.

The "job state flags" are gone from the public API and from the `flux job list` output.

There was some additional cleanup, mainly in the job manager.

I had these changes mixed in with the scheduler interface stuff, but thought it best to submit as its own PR.

I'll call it a WIP at this point because I think I can improve the testing a bit.

grondo commented:
This is nice! :+1:

garlick commented:
Just forced pushed a rebase on current master, the addition of some testing (mostly unit tests), and a couple of minor bug fixes uncovered in testing.

garlick commented:
I'm not sure what's going on with travis.  Looks like the tests have all passed but the little ball is still orange.  Hrm.

grondo commented:
I checked on travis and it says the status of your pr is "snowed in"

garlick commented:
That seems to have done the trick.

grondo commented on src/cmd/flux-job.c:
```diff
@@ -104,11 +108,18 @@ static struct optparse_subcommand subcommands[] = {
-      cmd_cancel,
+      "[OPTIONS] id [message ...]",
+      "Alias for flux job raise",
+      cmd_raise,
```
I think it is ok to hid the fact that `flux job cancel` is an alias for `flux job raise` and provide help message that says "Cancel a job" for better UI.

I love that you can provide a message when you cancel a job.

garlick replied:
Sounds good.

grondo commented on src/modules/job-manager/util.c:
```diff
@@ -0,0 +1,114 @@
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* active - manipulate active jobs in the KVS
```
minor nit: I assume this line is left over from when this file was `active.c`, should the comment read `util - manipulate active jobs in kvs` or similar?

garlick replied:
Oops! ! Yes sloppy error.

grondo commented on src/common/libjob/job.h:
```diff
@@ -74,9 +71,13 @@ int flux_job_submit_get_id (flux_future_t *f, flux_jobid_t *id);
+flux_future_t *flux_job_raise (flux_t *h, flux_jobid_t id,
+                               const char *type, int severity,
+                               const char *note);
 
```
Any reason not to keep a `flux_job_cancel` as an user-friendly alias for `flux_job_raise`? (I imagine an exception of type `cancel` will be over 90% of use cases, and saving users from writing `flux_job_raise (h, id, "cancel", 0, NULL)` will be a mercy)

Could even just be a macro at this point.

garlick replied:
OK, will do.

garlick commented:
Rebased on current master, fixed review comments, and pushed a couple of other minor little things that should be squashed down before this is merged.

grondo commented:
Looks like a builder hit #2015

garlick commented:
Great thanks!

On `flux job list` - JSON output is a good idea as it would let us write tests that are not dependent on column position of output.    I'll open an issue on that one.

garlick commented:
Rebased after @chu11's fix, and also squashed down some incremental development.

garlick commented:
Rebased.  (Thanks guys for the nice feedback!)

grondo commented:
Just needs a rebase and I think it is good to go in!

garlick commented:
Thanks, rebased and also snuck in a fix for this typo (squashed):

```diff
diff --git a/src/modules/job-manager/raise.c b/src/modules/job-manager/raise.c
index 861d4a5e..ffa6cea7 100644
--- a/src/modules/job-manager/raise.c
+++ b/src/modules/job-manager/raise.c
@@ -79,7 +79,7 @@ void raise_ctx_decref (struct raise_ctx *c)
 {
     if (c && --c->refcount == 0) {
         int saved_errno = errno;
-        c->job->flags &= JOB_EXCEPTION_PENDING;
+        c->job->flags &= ~JOB_EXCEPTION_PENDING;
         if (c->severity == 0)
             c->job->state = FLUX_JOB_CLEANUP;
         if (c->request) {
```

garlick commented:
Yesss!   Thanks.
